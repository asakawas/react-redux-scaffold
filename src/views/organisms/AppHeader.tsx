import React, { FunctionComponent } from 'react';
import Toolbar from '@material-ui/core/Toolbar';
import LeftMarginedAppBar, { LeftMarginedAppBarProps } from '~/views/atoms/LeftMarginedAppBar';
import HamburgerMenuButton, {
  HamburgerMenuButtonProps,
} from '~/views/atoms/buttons/HamburgerMenuButton';
import ZDKAccountMenu from '~/views/ecosystems/ZDKAccountMenu';
import ToolbarTitle from '~/views/atoms/ToolbarTitle';

/**
 * The type of props of AppHeader.
 */
export type AppHeaderProps = Readonly<
  HamburgerMenuButtonProps &
    Pick<LeftMarginedAppBarProps, 'marginLeft'> & {
      /** The title on the header. */
      title: string;

      /** Set true to show the menu button. */
      showMenuButton: boolean;
    }
>;

const AppHeader: FunctionComponent<AppHeaderProps> = ({
  title,
  onHamburgerMenuButtonClick,
  showMenuButton,
  marginLeft,
}) => (
  <LeftMarginedAppBar position="static" marginLeft={marginLeft} aria-label="app bar">
    <Toolbar>
      {showMenuButton ? (
        <HamburgerMenuButton onHamburgerMenuButtonClick={onHamburgerMenuButtonClick} />
      ) : (
        <></>
      )}
      <ToolbarTitle variant="h5">{title}</ToolbarTitle>
      <ZDKAccountMenu />
    </Toolbar>
  </LeftMarginedAppBar>
);

export default React.memo(AppHeader);
