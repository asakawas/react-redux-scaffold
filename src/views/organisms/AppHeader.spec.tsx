import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { render, cleanup, fireEvent } from '@testing-library/react';
import AppHeader from './AppHeader';

// We will use this empty store to wrap the testing component (that includes connected components)
// with Redux Provider to avoid an error at render().
const store = createStore(() => {});

jest.mock('~/state/ducks/app/selectors', () => ({
  getCurrentUser: () => ({ name: 'hoge' }),
}));

describe('AppHeader', () => {
  test('is rendered as expected', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <AppHeader
          title="Test"
          onHamburgerMenuButtonClick={() => {}}
          showMenuButton
          marginLeft={100}
        />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('shows the menu button only when showMenuButton is set to true', () => {
    {
      const renderResult = render(
        <Provider store={store}>
          <AppHeader
            title="Test"
            onHamburgerMenuButtonClick={() => {}}
            showMenuButton
            marginLeft={100}
          />
        </Provider>,
      );

      const menuButton = renderResult.queryByLabelText('hamburger menu');
      expect(menuButton).not.toBeNull();
    }

    cleanup();

    {
      const renderResult = render(
        <Provider store={store}>
          <AppHeader
            title="Test"
            onHamburgerMenuButtonClick={() => {}}
            showMenuButton={false}
            marginLeft={100}
          />
        </Provider>,
      );

      const menuButton = renderResult.queryByLabelText('hamburger menu');
      expect(menuButton).toBeNull();
    }
  });

  test('calls the passed handler when the menu button clicked', () => {
    const handler = jest.fn();
    const renderResult = render(
      <Provider store={store}>
        <AppHeader
          title="Test"
          onHamburgerMenuButtonClick={handler}
          showMenuButton
          marginLeft={100}
        />
      </Provider>,
    );

    const menuButton = renderResult.queryByLabelText('hamburger menu');
    fireEvent.click(menuButton);
    expect(handler).toHaveBeenCalledTimes(1);
    fireEvent.click(menuButton);
    expect(handler).toHaveBeenCalledTimes(2);
  });
});
