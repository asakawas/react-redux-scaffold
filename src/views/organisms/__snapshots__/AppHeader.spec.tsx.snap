// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader is rendered as expected 1`] = `
<DocumentFragment>
  .c0 {
  margin-left: 100px;
  width: calc(100% - 100px);
}

.c2 {
  background-color: #b3d6a4;
}

.c1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<header
    aria-label="app bar"
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary c0 MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        aria-label="hamburger menu"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeStart"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h5
        class="MuiTypography-root c1 MuiTypography-h5"
      >
        Test
      </h5>
      <button
        aria-label="account menu button"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circle c2 MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorAction"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</DocumentFragment>
`;
